### PR TITLE
plan: pushdown predicates made by groupby item on a aggregation

### PR DESCRIPTION
**Pull request by winoros**
When we do predicate pushdown, we can push down the predicate made by only one group-by item across the aggregation. Like "select s from (select sum(v2) as s, v1 from t group by v1) k where v1 > 1", it can be rewritten as "select s from (select sum(v2) as s, v1 from t where v1 > 1 group by v1) k ".
@hanfei1991 @zimulala @coocood @XuHuaiyu PTAL

### CODE REVIEW

**Comment by shenli** [comment]
@winoros Please sign the CLA.

File: plan/predicate_push_down.go [hunk 1]
```diff
@@ -20,6 +20,8 @@ import (
 	"github.com/pingcap/tidb/expression"
 	"github.com/pingcap/tidb/mysql"
 	"github.com/pingcap/tidb/util/types"
+	"strconv"
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Put system package in front of all.

File: plan/predicate_push_down.go [hunk 2]
```diff
@@ -564,16 +566,78 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
 }
 
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
-func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan, error) {
+func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) (ret []expression.Expression, retPlan LogicalPlan, err error) {
 	// TODO: implement aggregation push down.
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Remove this line.

File: plan/predicate_push_down.go [hunk 3]
```diff
@@ -564,16 +566,78 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+			}
+		}
+	}
 	var condsToPush []expression.Expression
 	for _, cond := range predicates {
-		if _, ok := cond.(*expression.Constant); ok {
+		switch v := cond.(type) {
+		case *expression.Constant:
 			condsToPush = append(condsToPush, cond)
+			ret = append(ret, cond)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Why append this to ret?

File: plan/predicate_push_down.go [hunk 4]
```diff
@@ -564,16 +566,78 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+		default:
+			ret = append(ret, cond)
 		}
 	}
 	p.baseLogicalPlan.PredicatePushDown(condsToPush)
-	return predicates, p, nil
+	return
+}
+
+func scalarFunctionByOneCol(sf *expression.ScalarFunction) (ok bool, col *expression.Column) {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: There is a function called "extractColumn" to implement this feature.......

File: plan/predicate_push_down.go [hunk 5]
```diff
@@ -564,16 +566,78 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+	}
 	var condsToPush []expression.Expression
 	for _, cond := range predicates {
-		if _, ok := cond.(*expression.Constant); ok {
+		switch v := cond.(type) {
+		case *expression.Constant:
 			condsToPush = append(condsToPush, cond)
+			ret = append(ret, cond)
+		case *expression.ScalarFunction:
+			if ok, col := scalarFunctionByOneCol(v); ok && col.IsAggOrSubq && colNameIsGroupByItem(col.String(), aggrID, colIDsInGroupBy) {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: If you want to check if the column is group by item, you can get the column's index id directly by p.schema.GetIndex(col).

File: plan/plan_test.go [hunk 6]
```diff
@@ -591,6 +591,21 @@ func (s *testPlanSuite) TestPredicatePushDown(c *C) {
 			first: "Join{Join{DataScan(t)->DataScan(t)}->DataScan(t)}->Projection",
 			best:  "Join{Join{DataScan(t)->DataScan(t)}->DataScan(t)}->Projection",
 		},
+		{
```
Comment (hanfei1991) [inline plan/plan_test.go]: add a test containing having clause.

File: plan/predicate_push_down.go [hunk 7]
```diff
@@ -564,16 +566,78 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
-func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan, error) {
+func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) (ret []expression.Expression, retPlan LogicalPlan, err error) {
 	// TODO: implement aggregation push down.
+	retPlan = p
+	colsOfGroupBy := p.GroupByItems
+	var aggrID, _ = strconv.Atoi(strings.Split(p.GetID(), "_")[1]) // form of Aggregation ID: aggregation_x
+	var colIDsInGroupBy []int
+	var exprs []expression.Expression
+	for id, aggrFun := range p.AggFuncs {
+		exprs = append(exprs, aggrFun.GetArgs()[0])
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Why not move this in front of line 580.

File: plan/predicate_push_down.go [hunk 8]
```diff
@@ -564,16 +564,55 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+			}
+		}
+	}
 	var condsToPush []expression.Expression
 	for _, cond := range predicates {
-		if _, ok := cond.(*expression.Constant); ok {
+		switch cond.(type) {
+		case *expression.Constant:
 			condsToPush = append(condsToPush, cond)
+			ret = append(ret, cond)
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: Add comment for this row.

File: plan/predicate_push_down.go [hunk 9]
```diff
@@ -563,17 +563,58 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
 
+// CheckIsGroupByItem check whether this column is a group-by item
+func (p *Aggregation) CheckIsGroupByItem(col *expression.Column) bool {
+	var tmpSchema expression.Schema
+	for id, fun := range p.AggFuncs {
+		if fun.GetName() != ast.AggFuncFirstRow {
+			continue
+		}
+		for _, item := range p.GroupByItems {
+			if _, ok := item.(*expression.Column); ok && item.Equal(fun.GetArgs()[0]) {
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: you don't need to find if a gby-col exits in agg funcs.

File: plan/predicate_push_down.go [hunk 10]
```diff
@@ -563,17 +563,58 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+	}
+	return tmpSchema.GetIndex(col) >= 0
+}
+
 // PredicatePushDown implements LogicalPlan PredicatePushDown interface.
-func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan, error) {
-	// TODO: implement aggregation push down.
+func (p *Aggregation) PredicatePushDown(predicates []expression.Expression) (ret []expression.Expression, retPlan LogicalPlan, err error) {
+	retPlan = p
+	var exprs []expression.Expression
```
Comment (hanfei1991) [inline plan/predicate_push_down.go]: change var name to make code more readable.

**Comment by hanfei1991** [comment]
Add test with multiple group by columns.

**Comment by hanfei1991** [comment]
rest LGTM

**Comment by hanfei1991** [comment]
@coocood @XuHuaiyu @zimulala @shenli PTAL

**Comment by hanfei1991** [comment]
LGTM

File: plan/logical_plans.go [hunk 11]
```diff
@@ -66,6 +66,9 @@ type Aggregation struct {
 	AggFuncs     []expression.AggregationFunction
 	GroupByItems []expression.Expression
 	ctx          context.Context
+
+	// groupByCols stores the columns that are group-by items
+	groupByCols []*expression.Column
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: 1. add period('.') at the end of this comment.
2. align the format.

File: plan/predicate_push_down.go [hunk 12]
```diff
@@ -563,17 +563,49 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
 	return
 }
 
+// CheckColumnIsGroupByItem check whether a column is a group-by item
```
Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: add a period.

File: plan/predicate_push_down.go [hunk 13]
```diff
@@ -563,17 +563,49 @@ func (p *Union) PredicatePushDown(predicates []expression.Expression) (ret []exp
...
+			extractedCols, _ := extractColumn(cond, nil, nil)
+			ok := true
+			for _, col := range extractedCols {
+				if !p.CheckColumnIsGroupByItem(col) {
+					ok = false
+					break
+				}
+			}
+			if ok {
+				newFunc := columnSubstitute(cond.Clone(), p.GetSchema(), exprsOriginal)
```
Comment (XuHuaiyu) [inline plan/predicate_push_down.go]: can we pass `cond` to `columnSubstitute()` directly without Clone()?

File: plan/logical_plan_builder.go [hunk 14]
```diff
@@ -81,6 +89,7 @@ func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.Aggrega
 	}
 	agg.GroupByItems = gby
 	agg.SetSchema(schema)
+	agg.buildGroupByColumns()
```
Comment (XuHuaiyu) [inline plan/logical_plan_builder.go]: will `collectGroupByColumns` be better?

File: plan/predicate_push_down.go (hunk 13 above)
Comment (hanfei1991) [inline plan/predicate_push_down.go]: no, I will cause a panic. Some columns share same pointer.

**Comment by XuHuaiyu** [comment]
Maybe a test case for the situation that can not be pushed down.

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**